src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java

```diff
@@ -22,101 +22,105 @@
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.REQUEST)
-            .count(0)
-            .build();
-    final InstructionCount cancelCount =
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.CANCEL_REQUEST)
-            .count(0)
-            .build();
-
+  private void processActionType(final ActionType actionType) {
+    String actionTypeName = actionType.getName();
+    RLock lock = redissonClient.getFairLock(lockPrefix + actionTypeName);
     try {
-      final List<Action> actions = retrieveActions(actionType);
-      processActions(actions, requestCount, cancelCount);
-    } catch (final Exception e) {
-      log.with("action_type", actionType.getName()).error("Failed to process action type", e);
+      // Wait for a lock. Automatically unlock after a certain amount of time to prevent issues
+      // when lock holder crashes or Redis crashes causing permanent lockout
+      if (lock.tryLock(
+          appConfig.getDataGrid().getLockTimeToWaitSeconds(),
+          appConfig.getDataGrid().getLockTimeToLiveSeconds(),
+          TimeUnit.SECONDS)) {
+        try {
+          List<Action> actions =
+              actionRepo.findSubmittedOrCancelledByActionTypeName(
+                  actionType.getName(), appConfig.getActionDistribution().getRetrievalMax());
+          actions.forEach(this::processAction);
+        } finally {
+          // Always unlock the distributed lock
+          lock.unlock();
+        }
+      }
+    } catch (InterruptedException ex) {
+      // Ignored - process stopped while waiting for lock
     }
-    return Arrays.asList(requestCount, cancelCount);
   }
 
-  private void processActions(
-      final List<Action> actions,
-      final InstructionCount requestCount,
-      final InstructionCount cancelCount) {
-    log.with(actions).debug("Dealing with actions");
-    for (final Action action : actions) {
-      try {
-        processAction(action, requestCount, cancelCount);
-      } catch (final Exception e) {
-        log.with("action_id", action.getId())
-            .error("Could not process action, will be retried at next scheduled distribution", e);
+  private void processAction(Action action) {
+    ActionProcessingService ap = getActionProcessingService(action);
+
+    // If social reminder action type then generate new IAC
+    if (action.getActionType().getActionTypeNameEnum()
+        == uk.gov.ons.ctp.response.action.representation.ActionType.SOCIALREM) {
+      caseSvcClientService.generateNewIacForCase(action.getCaseId());
+    }
+
+    try {
+      if (action.getState().equals(ActionState.SUBMITTED)) {
+        ap.processActionRequests(action);
+      } else if (action.getState().equals(ActionState.CANCEL_SUBMITTED)) {
+        ap.processActionCancel(action);
       }
+    } catch (CTPException ex) {
```

Comment:
I hate CTPException so much. I actually catch it at the point where it's thrown and re-throw it as something useful like IllegalStateException, so we don't have to pollute our code with loads of useless `throws` declarations and `try...catch`es.